// angular-rave/bundles/angular-rave.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('angular-rave', ['exports', '@angular/core', '@angular/common'], factory) :
    (global = global || self, factory(global['angular-rave'] = {}, global.ng.core, global.ng.common));
}(this, function (exports, core, common) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (_) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    function __exportStar(m, exports) {
        for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
    }

    function __values(o) {
        var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
        if (m) return m.call(o);
        return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    };

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
        result.default = mod;
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @type {?} */
    var PBFPUBKEY_TOKEN = new core.InjectionToken('Rave.PBFPubKey');
    /** @type {?} */
    var ENVIRONMENT_TOKEN = new core.InjectionToken('Rave.IsDevelopment');

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * @record
     */
    function MyWindow() { }
    if (false) {
        /** @type {?} */
        MyWindow.prototype.getpaidSetup;
    }
    /** @type {?} */
    var PROD_URL = '//api.ravepay.co/flwv3-pug/getpaidx/api/flwpbf-inline.js';
    /** @type {?} */
    var DEV_URL = '//ravesandboxapi.flutterwave.com/flwv3-pug/getpaidx/api/flwpbf-inline.js';
    var AngularRaveService = /** @class */ (function () {
        function AngularRaveService(PBFPubKey, isDev) {
            this.PBFPubKey = PBFPubKey;
            this.isDev = isDev;
        }
        /**
         * @param {?} obj
         * @return {?}
         */
        AngularRaveService.prototype.createRaveOptionsObject = /**
         * @param {?} obj
         * @return {?}
         */
        function (obj) {
            /** @type {?} */
            var raveOptions = {};
            raveOptions.amount = obj.amount;
            // raveOptions.hosted_payment = 1; // Modal should open in another page
            raveOptions.PBFPubKey = obj.PBFPubKey || this.PBFPubKey;
            if (obj.payment_method) {
                raveOptions.payment_method = obj.payment_method;
            }
            if (obj.redirect_url) {
                raveOptions.redirect_url = obj.redirect_url;
            }
            if (obj.integrity_hash) {
                raveOptions.integrity_hash = obj.integrity_hash;
            }
            if (obj.pay_button_text) {
                raveOptions.pay_button_text = obj.pay_button_text;
            }
            if (obj.country) {
                raveOptions.country = obj.country;
            }
            if (obj.currency) {
                raveOptions.currency = obj.currency;
            }
            if (obj.custom_description) {
                raveOptions.custom_description = obj.custom_description;
            }
            if (obj.customer_email) {
                raveOptions.customer_email = obj.customer_email;
            }
            if (obj.custom_logo) {
                raveOptions.custom_logo = obj.custom_logo;
            }
            if (obj.custom_title) {
                raveOptions.custom_title = obj.custom_title;
            }
            if (obj.customer_firstname) {
                raveOptions.customer_firstname = obj.customer_firstname;
            }
            if (obj.subaccount) {
                raveOptions.subaccount = obj.subaccount;
            }
            if (obj.customer_lastname) {
                raveOptions.customer_lastname = obj.customer_lastname;
            }
            if (obj.txref) {
                raveOptions.txref = obj.txref;
            }
            if (obj.customer_phone) {
                raveOptions.customer_phone = obj.customer_phone;
            }
            return raveOptions;
        };
        /**
         * @return {?}
         */
        AngularRaveService.prototype.loadScript = /**
         * @return {?}
         */
        function () {
            var _this = this;
            return new Promise((/**
             * @param {?} resolve
             * @return {?}
             */
            function (resolve) {
                if (typeof window.getpaidSetup === 'function') {
                    resolve();
                    return;
                }
                /** @type {?} */
                var script = window.document.createElement('script');
                window.document.head.appendChild(script);
                /** @type {?} */
                var onLoadFunc = (/**
                 * @return {?}
                 */
                function () {
                    script.removeEventListener('load', onLoadFunc);
                    resolve();
                });
                script.addEventListener('load', onLoadFunc);
                /** @type {?} */
                var url = _this.isDev ? DEV_URL : PROD_URL;
                script.setAttribute('src', url);
            }));
        };
        /**
         * @param {?} obj
         * @return {?}
         */
        AngularRaveService.prototype.isInvalidOptions = /**
         * @param {?} obj
         * @return {?}
         */
        function (obj) {
            if (!obj.PBFPubKey && !this.PBFPubKey) {
                return 'ANGULAR-RAVE: Merchant public key is required';
            }
            if (!(obj.customer_email || obj.customer_phone)) {
                return 'ANGULAR-RAVE: Customer email or phone number is required';
            }
            if (!obj.txref) {
                return 'ANGULAR-RAVE: A unique transaction reference is required';
            }
            if (!obj.currency) {
                return 'ANGULAR-RAVE: Currency is required, Use "NGN" for naira';
            }
            if (!obj.amount) {
                return 'ANGULAR-RAVE: Amount to charge is required';
            }
            return '';
        };
        AngularRaveService.decorators = [
            { type: core.Injectable, args: [{
                        providedIn: 'root',
                    },] }
        ];
        /** @nocollapse */
        AngularRaveService.ctorParameters = function () { return [
            { type: String, decorators: [{ type: core.Inject, args: [PBFPUBKEY_TOKEN,] }] },
            { type: Boolean, decorators: [{ type: core.Inject, args: [ENVIRONMENT_TOKEN,] }] }
        ]; };
        /** @nocollapse */ AngularRaveService.ngInjectableDef = core.ɵɵdefineInjectable({ factory: function AngularRaveService_Factory() { return new AngularRaveService(core.ɵɵinject(PBFPUBKEY_TOKEN), core.ɵɵinject(ENVIRONMENT_TOKEN)); }, token: AngularRaveService, providedIn: "root" });
        return AngularRaveService;
    }());
    if (false) {
        /**
         * @type {?}
         * @private
         */
        AngularRaveService.prototype.PBFPubKey;
        /**
         * @type {?}
         * @private
         */
        AngularRaveService.prototype.isDev;
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * @record
     */
    function MyWindow$1() { }
    if (false) {
        /** @type {?} */
        MyWindow$1.prototype.getpaidSetup;
    }
    var AngularRaveComponent = /** @class */ (function () {
        function AngularRaveComponent(raveService) {
            this.raveService = raveService;
            this.onclose = new core.EventEmitter();
            this.callback = new core.EventEmitter();
            this.init = new core.EventEmitter();
            this._raveOptions = {};
        }
        /**
         * @return {?}
         */
        AngularRaveComponent.prototype.pay = /**
         * @return {?}
         */
        function () {
            return __awaiter(this, void 0, void 0, function () {
                var errorExists;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            errorExists = false;
                            if (this.raveOptions && Object.keys(this.raveOptions).length > 1) {
                                errorExists = this.checkInvalidOptions(this.raveOptions);
                                this.insertRaveOptions(this.raveOptions);
                            }
                            else {
                                errorExists = this.checkInvalidOptions(this);
                                this.insertRaveOptions(this);
                            }
                            if (errorExists) {
                                return [2 /*return*/];
                            }
                            return [4 /*yield*/, this.raveService.loadScript()];
                        case 1:
                            _a.sent();
                            this.paymentSetup = window.getpaidSetup(this._raveOptions);
                            if (this.init.observers.length > 0) {
                                this.init.emit(this.paymentSetup);
                            }
                            return [2 /*return*/];
                    }
                });
            });
        };
        /**
         * @param {?} object
         * @return {?}
         */
        AngularRaveComponent.prototype.checkInvalidOptions = /**
         * @param {?} object
         * @return {?}
         */
        function (object) {
            /** @type {?} */
            var optionsInvalid = this.raveService.isInvalidOptions(object);
            if (optionsInvalid) {
                console.error(optionsInvalid);
            }
            return optionsInvalid !== '';
        };
        /**
         * @param {?} object
         * @return {?}
         */
        AngularRaveComponent.prototype.insertRaveOptions = /**
         * @param {?} object
         * @return {?}
         */
        function (object) {
            var _this = this;
            this._raveOptions = this.raveService.createRaveOptionsObject(object);
            if (this.onclose) {
                this._raveOptions.onclose = (/**
                 * @return {?}
                 */
                function () { return _this.onclose.emit(); });
            }
            this._raveOptions.callback = (/**
             * @param {?} res
             * @return {?}
             */
            function (res) {
                _this.onclose.emit(res);
                if (_this.autoClose) {
                    _this.paymentSetup.close();
                }
            });
        };
        /**
         * @return {?}
         */
        AngularRaveComponent.prototype.ngOnInit = /**
         * @return {?}
         */
        function () {
            this.pay();
        };
        AngularRaveComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'angular-rave',
                        // tslint:disable-line
                        template: "<ng-content></ng-content>"
                    }] }
        ];
        /** @nocollapse */
        AngularRaveComponent.ctorParameters = function () { return [
            { type: AngularRaveService }
        ]; };
        AngularRaveComponent.propDecorators = {
            amount: [{ type: core.Input }],
            autoClose: [{ type: core.Input }],
            country: [{ type: core.Input }],
            currency: [{ type: core.Input }],
            custom_description: [{ type: core.Input }],
            custom_logo: [{ type: core.Input }],
            custom_title: [{ type: core.Input }],
            customer_email: [{ type: core.Input }],
            customer_firstname: [{ type: core.Input }],
            customer_lastname: [{ type: core.Input }],
            customer_phone: [{ type: core.Input }],
            integrity_hash: [{ type: core.Input }],
            meta: [{ type: core.Input }],
            pay_button_text: [{ type: core.Input }],
            payment_method: [{ type: core.Input }],
            PBFPubKey: [{ type: core.Input }],
            raveOptions: [{ type: core.Input }],
            redirect_url: [{ type: core.Input }],
            subaccount: [{ type: core.Input }],
            txref: [{ type: core.Input }],
            onclose: [{ type: core.Output }],
            callback: [{ type: core.Output }],
            init: [{ type: core.Output }]
        };
        return AngularRaveComponent;
    }());
    if (false) {
        /** @type {?} */
        AngularRaveComponent.prototype.amount;
        /** @type {?} */
        AngularRaveComponent.prototype.autoClose;
        /** @type {?} */
        AngularRaveComponent.prototype.country;
        /** @type {?} */
        AngularRaveComponent.prototype.currency;
        /** @type {?} */
        AngularRaveComponent.prototype.custom_description;
        /** @type {?} */
        AngularRaveComponent.prototype.custom_logo;
        /** @type {?} */
        AngularRaveComponent.prototype.custom_title;
        /** @type {?} */
        AngularRaveComponent.prototype.customer_email;
        /** @type {?} */
        AngularRaveComponent.prototype.customer_firstname;
        /** @type {?} */
        AngularRaveComponent.prototype.customer_lastname;
        /** @type {?} */
        AngularRaveComponent.prototype.customer_phone;
        /** @type {?} */
        AngularRaveComponent.prototype.integrity_hash;
        /** @type {?} */
        AngularRaveComponent.prototype.meta;
        /** @type {?} */
        AngularRaveComponent.prototype.pay_button_text;
        /** @type {?} */
        AngularRaveComponent.prototype.payment_method;
        /** @type {?} */
        AngularRaveComponent.prototype.PBFPubKey;
        /** @type {?} */
        AngularRaveComponent.prototype.raveOptions;
        /** @type {?} */
        AngularRaveComponent.prototype.redirect_url;
        /** @type {?} */
        AngularRaveComponent.prototype.subaccount;
        /** @type {?} */
        AngularRaveComponent.prototype.txref;
        /** @type {?} */
        AngularRaveComponent.prototype.onclose;
        /** @type {?} */
        AngularRaveComponent.prototype.callback;
        /** @type {?} */
        AngularRaveComponent.prototype.init;
        /**
         * @type {?}
         * @private
         */
        AngularRaveComponent.prototype._raveOptions;
        /**
         * @type {?}
         * @private
         */
        AngularRaveComponent.prototype.paymentSetup;
        /**
         * @type {?}
         * @private
         */
        AngularRaveComponent.prototype.raveService;
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * @record
     */
    function MyWindow$2() { }
    if (false) {
        /** @type {?} */
        MyWindow$2.prototype.getpaidSetup;
    }
    var AngularRaveDirective = /** @class */ (function () {
        function AngularRaveDirective(raveService) {
            this.raveService = raveService;
            this.raveOptions = {};
            this.onclose = new core.EventEmitter();
            this.callback = new core.EventEmitter();
            this.init = new core.EventEmitter();
            this._raveOptions = {};
        }
        /**
         * @return {?}
         */
        AngularRaveDirective.prototype.buttonClick = /**
         * @return {?}
         */
        function () {
            this.pay();
        };
        /**
         * @return {?}
         */
        AngularRaveDirective.prototype.pay = /**
         * @return {?}
         */
        function () {
            return __awaiter(this, void 0, void 0, function () {
                var errorExists;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            errorExists = false;
                            if (this.raveOptions && Object.keys(this.raveOptions).length > 1) {
                                errorExists = this.checkInvalidOptions(this.raveOptions);
                                this.insertRaveOptions(this.raveOptions);
                            }
                            else {
                                errorExists = this.checkInvalidOptions(this);
                                this.insertRaveOptions(this);
                            }
                            if (errorExists) {
                                return [2 /*return*/];
                            }
                            return [4 /*yield*/, this.raveService.loadScript()];
                        case 1:
                            _a.sent();
                            this.paymentSetup = window.getpaidSetup(this._raveOptions);
                            if (this.init.observers.length > 0) {
                                this.init.emit(this.paymentSetup);
                            }
                            return [2 /*return*/];
                    }
                });
            });
        };
        /**
         * @param {?} object
         * @return {?}
         */
        AngularRaveDirective.prototype.checkInvalidOptions = /**
         * @param {?} object
         * @return {?}
         */
        function (object) {
            /** @type {?} */
            var optionsInvalid = this.raveService.isInvalidOptions(object);
            if (optionsInvalid) {
                console.error(optionsInvalid);
            }
            return optionsInvalid !== '';
        };
        /**
         * @param {?} object
         * @return {?}
         */
        AngularRaveDirective.prototype.insertRaveOptions = /**
         * @param {?} object
         * @return {?}
         */
        function (object) {
            var _this = this;
            this._raveOptions = this.raveService.createRaveOptionsObject(object);
            if (this.onclose) {
                this._raveOptions.onclose = (/**
                 * @return {?}
                 */
                function () { return _this.onclose.emit(); });
            }
            this._raveOptions.callback = (/**
             * @param {?} res
             * @return {?}
             */
            function (res) {
                _this.onclose.emit(res);
                if (_this.autoClose) {
                    _this.paymentSetup.close();
                }
            });
        };
        AngularRaveDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: '[angular-rave]',
                    },] }
        ];
        /** @nocollapse */
        AngularRaveDirective.ctorParameters = function () { return [
            { type: AngularRaveService }
        ]; };
        AngularRaveDirective.propDecorators = {
            amount: [{ type: core.Input }],
            autoClose: [{ type: core.Input }],
            country: [{ type: core.Input }],
            currency: [{ type: core.Input }],
            custom_description: [{ type: core.Input }],
            custom_logo: [{ type: core.Input }],
            custom_title: [{ type: core.Input }],
            customer_email: [{ type: core.Input }],
            customer_firstname: [{ type: core.Input }],
            customer_lastname: [{ type: core.Input }],
            customer_phone: [{ type: core.Input }],
            integrity_hash: [{ type: core.Input }],
            meta: [{ type: core.Input }],
            pay_button_text: [{ type: core.Input }],
            payment_method: [{ type: core.Input }],
            PBFPubKey: [{ type: core.Input }],
            raveOptions: [{ type: core.Input }],
            redirect_url: [{ type: core.Input }],
            subaccount: [{ type: core.Input }],
            txref: [{ type: core.Input }],
            onclose: [{ type: core.Output }],
            callback: [{ type: core.Output }],
            init: [{ type: core.Output }],
            buttonClick: [{ type: core.HostListener, args: ['click',] }]
        };
        return AngularRaveDirective;
    }());
    if (false) {
        /** @type {?} */
        AngularRaveDirective.prototype.amount;
        /** @type {?} */
        AngularRaveDirective.prototype.autoClose;
        /** @type {?} */
        AngularRaveDirective.prototype.country;
        /** @type {?} */
        AngularRaveDirective.prototype.currency;
        /** @type {?} */
        AngularRaveDirective.prototype.custom_description;
        /** @type {?} */
        AngularRaveDirective.prototype.custom_logo;
        /** @type {?} */
        AngularRaveDirective.prototype.custom_title;
        /** @type {?} */
        AngularRaveDirective.prototype.customer_email;
        /** @type {?} */
        AngularRaveDirective.prototype.customer_firstname;
        /** @type {?} */
        AngularRaveDirective.prototype.customer_lastname;
        /** @type {?} */
        AngularRaveDirective.prototype.customer_phone;
        /** @type {?} */
        AngularRaveDirective.prototype.integrity_hash;
        /** @type {?} */
        AngularRaveDirective.prototype.meta;
        /** @type {?} */
        AngularRaveDirective.prototype.pay_button_text;
        /** @type {?} */
        AngularRaveDirective.prototype.payment_method;
        /** @type {?} */
        AngularRaveDirective.prototype.PBFPubKey;
        /** @type {?} */
        AngularRaveDirective.prototype.raveOptions;
        /** @type {?} */
        AngularRaveDirective.prototype.redirect_url;
        /** @type {?} */
        AngularRaveDirective.prototype.subaccount;
        /** @type {?} */
        AngularRaveDirective.prototype.txref;
        /** @type {?} */
        AngularRaveDirective.prototype.onclose;
        /** @type {?} */
        AngularRaveDirective.prototype.callback;
        /** @type {?} */
        AngularRaveDirective.prototype.init;
        /**
         * @type {?}
         * @private
         */
        AngularRaveDirective.prototype._raveOptions;
        /**
         * @type {?}
         * @private
         */
        AngularRaveDirective.prototype.paymentSetup;
        /**
         * @type {?}
         * @private
         */
        AngularRaveDirective.prototype.raveService;
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var AngularRaveModule = /** @class */ (function () {
        function AngularRaveModule() {
        }
        /**
         * @param {?} options
         * @return {?}
         */
        AngularRaveModule.forRoot = /**
         * @param {?} options
         * @return {?}
         */
        function (options) {
            return {
                ngModule: AngularRaveModule,
                providers: [
                    AngularRaveService,
                    { provide: PBFPUBKEY_TOKEN, useValue: options.key },
                    { provide: ENVIRONMENT_TOKEN, useValue: options.isTest }
                ]
            };
        };
        AngularRaveModule.decorators = [
            { type: core.NgModule, args: [{
                        imports: [common.CommonModule],
                        declarations: [AngularRaveDirective, AngularRaveComponent],
                        exports: [AngularRaveDirective, AngularRaveComponent],
                    },] }
        ];
        return AngularRaveModule;
    }());

    exports.AngularRaveComponent = AngularRaveComponent;
    exports.AngularRaveDirective = AngularRaveDirective;
    exports.AngularRaveModule = AngularRaveModule;
    exports.AngularRaveService = AngularRaveService;
    exports.ɵa = PBFPUBKEY_TOKEN;
    exports.ɵb = ENVIRONMENT_TOKEN;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
