// angular-rave/esm2015/lib/angular-rave.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Inject } from '@angular/core';
import { PBFPUBKEY_TOKEN, ENVIRONMENT_TOKEN } from './angular-rave-token';
import * as i0 from "@angular/core";
import * as i1 from "./angular-rave-token";
/**
 * @record
 */
function MyWindow() { }
if (false) {
    /** @type {?} */
    MyWindow.prototype.getpaidSetup;
}
/** @type {?} */
const PROD_URL = '//api.ravepay.co/flwv3-pug/getpaidx/api/flwpbf-inline.js';
/** @type {?} */
const DEV_URL = '//ravesandboxapi.flutterwave.com/flwv3-pug/getpaidx/api/flwpbf-inline.js';
export class AngularRaveService {
    /**
     * @param {?} PBFPubKey
     * @param {?} isDev
     */
    constructor(PBFPubKey, isDev) {
        this.PBFPubKey = PBFPubKey;
        this.isDev = isDev;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    createRaveOptionsObject(obj) {
        /** @type {?} */
        const raveOptions = {};
        raveOptions.amount = obj.amount;
        // raveOptions.hosted_payment = 1; // Modal should open in another page
        raveOptions.PBFPubKey = obj.PBFPubKey || this.PBFPubKey;
        if (obj.payment_method) {
            raveOptions.payment_method = obj.payment_method;
        }
        if (obj.redirect_url) {
            raveOptions.redirect_url = obj.redirect_url;
        }
        if (obj.integrity_hash) {
            raveOptions.integrity_hash = obj.integrity_hash;
        }
        if (obj.pay_button_text) {
            raveOptions.pay_button_text = obj.pay_button_text;
        }
        if (obj.country) {
            raveOptions.country = obj.country;
        }
        if (obj.currency) {
            raveOptions.currency = obj.currency;
        }
        if (obj.custom_description) {
            raveOptions.custom_description = obj.custom_description;
        }
        if (obj.customer_email) {
            raveOptions.customer_email = obj.customer_email;
        }
        if (obj.custom_logo) {
            raveOptions.custom_logo = obj.custom_logo;
        }
        if (obj.custom_title) {
            raveOptions.custom_title = obj.custom_title;
        }
        if (obj.customer_firstname) {
            raveOptions.customer_firstname = obj.customer_firstname;
        }
        if (obj.subaccount) {
            raveOptions.subaccount = obj.subaccount;
        }
        if (obj.customer_lastname) {
            raveOptions.customer_lastname = obj.customer_lastname;
        }
        if (obj.txref) {
            raveOptions.txref = obj.txref;
        }
        if (obj.customer_phone) {
            raveOptions.customer_phone = obj.customer_phone;
        }
        return raveOptions;
    }
    /**
     * @return {?}
     */
    loadScript() {
        return new Promise((/**
         * @param {?} resolve
         * @return {?}
         */
        resolve => {
            if (typeof window.getpaidSetup === 'function') {
                resolve();
                return;
            }
            /** @type {?} */
            const script = window.document.createElement('script');
            window.document.head.appendChild(script);
            /** @type {?} */
            const onLoadFunc = (/**
             * @return {?}
             */
            () => {
                script.removeEventListener('load', onLoadFunc);
                resolve();
            });
            script.addEventListener('load', onLoadFunc);
            /** @type {?} */
            const url = this.isDev ? DEV_URL : PROD_URL;
            script.setAttribute('src', url);
        }));
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    isInvalidOptions(obj) {
        if (!obj.PBFPubKey && !this.PBFPubKey) {
            return 'ANGULAR-RAVE: Merchant public key is required';
        }
        if (!(obj.customer_email || obj.customer_phone)) {
            return 'ANGULAR-RAVE: Customer email or phone number is required';
        }
        if (!obj.txref) {
            return 'ANGULAR-RAVE: A unique transaction reference is required';
        }
        if (!obj.currency) {
            return 'ANGULAR-RAVE: Currency is required, Use "NGN" for naira';
        }
        if (!obj.amount) {
            return 'ANGULAR-RAVE: Amount to charge is required';
        }
        return '';
    }
}
AngularRaveService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
/** @nocollapse */
AngularRaveService.ctorParameters = () => [
    { type: String, decorators: [{ type: Inject, args: [PBFPUBKEY_TOKEN,] }] },
    { type: Boolean, decorators: [{ type: Inject, args: [ENVIRONMENT_TOKEN,] }] }
];
/** @nocollapse */ AngularRaveService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function AngularRaveService_Factory() { return new AngularRaveService(i0.ɵɵinject(i1.PBFPUBKEY_TOKEN), i0.ɵɵinject(i1.ENVIRONMENT_TOKEN)); }, token: AngularRaveService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    AngularRaveService.prototype.PBFPubKey;
    /**
     * @type {?}
     * @private
     */
    AngularRaveService.prototype.isDev;
}
