// angular-rave/esm5/lib/angular-rave.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AngularRaveComponent } from './angular-rave.component';
import { AngularRaveDirective } from './angular-rave.directive';
import { AngularRaveService } from './angular-rave.service';
import { PBFPUBKEY_TOKEN, ENVIRONMENT_TOKEN } from './angular-rave-token';
var AngularRaveModule = /** @class */ (function () {
    function AngularRaveModule() {
    }
    /**
     * @param {?} options
     * @return {?}
     */
    AngularRaveModule.forRoot = /**
     * @param {?} options
     * @return {?}
     */
    function (options) {
        return {
            ngModule: AngularRaveModule,
            providers: [
                AngularRaveService,
                { provide: PBFPUBKEY_TOKEN, useValue: options.key },
                { provide: ENVIRONMENT_TOKEN, useValue: options.isTest }
            ]
        };
    };
    AngularRaveModule.decorators = [
        { type: NgModule, args: [{
                    imports: [CommonModule],
                    declarations: [AngularRaveDirective, AngularRaveComponent],
                    exports: [AngularRaveDirective, AngularRaveComponent],
                },] }
    ];
    return AngularRaveModule;
}());
export { AngularRaveModule };
