// angular-rave/esm5/lib/angular-rave.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Directive, Input, Output, HostListener, EventEmitter } from '@angular/core';
import { AngularRaveService } from './angular-rave.service';
/**
 * @record
 */
function MyWindow() { }
if (false) {
    /** @type {?} */
    MyWindow.prototype.getpaidSetup;
}
var AngularRaveDirective = /** @class */ (function () {
    function AngularRaveDirective(raveService) {
        this.raveService = raveService;
        this.raveOptions = {};
        this.onclose = new EventEmitter();
        this.callback = new EventEmitter();
        this.init = new EventEmitter();
        this._raveOptions = {};
    }
    /**
     * @return {?}
     */
    AngularRaveDirective.prototype.buttonClick = /**
     * @return {?}
     */
    function () {
        this.pay();
    };
    /**
     * @return {?}
     */
    AngularRaveDirective.prototype.pay = /**
     * @return {?}
     */
    function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var errorExists;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        errorExists = false;
                        if (this.raveOptions && Object.keys(this.raveOptions).length > 1) {
                            errorExists = this.checkInvalidOptions(this.raveOptions);
                            this.insertRaveOptions(this.raveOptions);
                        }
                        else {
                            errorExists = this.checkInvalidOptions(this);
                            this.insertRaveOptions(this);
                        }
                        if (errorExists) {
                            return [2 /*return*/];
                        }
                        return [4 /*yield*/, this.raveService.loadScript()];
                    case 1:
                        _a.sent();
                        this.paymentSetup = window.getpaidSetup(this._raveOptions);
                        if (this.init.observers.length > 0) {
                            this.init.emit(this.paymentSetup);
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * @param {?} object
     * @return {?}
     */
    AngularRaveDirective.prototype.checkInvalidOptions = /**
     * @param {?} object
     * @return {?}
     */
    function (object) {
        /** @type {?} */
        var optionsInvalid = this.raveService.isInvalidOptions(object);
        if (optionsInvalid) {
            console.error(optionsInvalid);
        }
        return optionsInvalid !== '';
    };
    /**
     * @param {?} object
     * @return {?}
     */
    AngularRaveDirective.prototype.insertRaveOptions = /**
     * @param {?} object
     * @return {?}
     */
    function (object) {
        var _this = this;
        this._raveOptions = this.raveService.createRaveOptionsObject(object);
        if (this.onclose) {
            this._raveOptions.onclose = (/**
             * @return {?}
             */
            function () { return _this.onclose.emit(); });
        }
        this._raveOptions.callback = (/**
         * @param {?} res
         * @return {?}
         */
        function (res) {
            _this.onclose.emit(res);
            if (_this.autoClose) {
                _this.paymentSetup.close();
            }
        });
    };
    AngularRaveDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[angular-rave]',
                },] }
    ];
    /** @nocollapse */
    AngularRaveDirective.ctorParameters = function () { return [
        { type: AngularRaveService }
    ]; };
    AngularRaveDirective.propDecorators = {
        amount: [{ type: Input }],
        autoClose: [{ type: Input }],
        country: [{ type: Input }],
        currency: [{ type: Input }],
        custom_description: [{ type: Input }],
        custom_logo: [{ type: Input }],
        custom_title: [{ type: Input }],
        customer_email: [{ type: Input }],
        customer_firstname: [{ type: Input }],
        customer_lastname: [{ type: Input }],
        customer_phone: [{ type: Input }],
        integrity_hash: [{ type: Input }],
        meta: [{ type: Input }],
        pay_button_text: [{ type: Input }],
        payment_method: [{ type: Input }],
        PBFPubKey: [{ type: Input }],
        raveOptions: [{ type: Input }],
        redirect_url: [{ type: Input }],
        subaccount: [{ type: Input }],
        txref: [{ type: Input }],
        onclose: [{ type: Output }],
        callback: [{ type: Output }],
        init: [{ type: Output }],
        buttonClick: [{ type: HostListener, args: ['click',] }]
    };
    return AngularRaveDirective;
}());
export { AngularRaveDirective };
if (false) {
    /** @type {?} */
    AngularRaveDirective.prototype.amount;
    /** @type {?} */
    AngularRaveDirective.prototype.autoClose;
    /** @type {?} */
    AngularRaveDirective.prototype.country;
    /** @type {?} */
    AngularRaveDirective.prototype.currency;
    /** @type {?} */
    AngularRaveDirective.prototype.custom_description;
    /** @type {?} */
    AngularRaveDirective.prototype.custom_logo;
    /** @type {?} */
    AngularRaveDirective.prototype.custom_title;
    /** @type {?} */
    AngularRaveDirective.prototype.customer_email;
    /** @type {?} */
    AngularRaveDirective.prototype.customer_firstname;
    /** @type {?} */
    AngularRaveDirective.prototype.customer_lastname;
    /** @type {?} */
    AngularRaveDirective.prototype.customer_phone;
    /** @type {?} */
    AngularRaveDirective.prototype.integrity_hash;
    /** @type {?} */
    AngularRaveDirective.prototype.meta;
    /** @type {?} */
    AngularRaveDirective.prototype.pay_button_text;
    /** @type {?} */
    AngularRaveDirective.prototype.payment_method;
    /** @type {?} */
    AngularRaveDirective.prototype.PBFPubKey;
    /** @type {?} */
    AngularRaveDirective.prototype.raveOptions;
    /** @type {?} */
    AngularRaveDirective.prototype.redirect_url;
    /** @type {?} */
    AngularRaveDirective.prototype.subaccount;
    /** @type {?} */
    AngularRaveDirective.prototype.txref;
    /** @type {?} */
    AngularRaveDirective.prototype.onclose;
    /** @type {?} */
    AngularRaveDirective.prototype.callback;
    /** @type {?} */
    AngularRaveDirective.prototype.init;
    /**
     * @type {?}
     * @private
     */
    AngularRaveDirective.prototype._raveOptions;
    /**
     * @type {?}
     * @private
     */
    AngularRaveDirective.prototype.paymentSetup;
    /**
     * @type {?}
     * @private
     */
    AngularRaveDirective.prototype.raveService;
}
