// angular-rave/esm5/lib/rave-options.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function RaveOptions() { }
if (false) {
    /**
     * Your merchant public key provided when you create a button
     * @type {?|undefined}
     */
    RaveOptions.prototype.PBFPubKey;
    /**
     * Email of the customer
     * @type {?|undefined}
     */
    RaveOptions.prototype.customer_email;
    /**
     * Firstname of the customer
     * @type {?|undefined}
     */
    RaveOptions.prototype.customer_firstname;
    /**
     * firstname of the customer
     * @type {?|undefined}
     */
    RaveOptions.prototype.customer_lastname;
    /**
     * Text to be displayed as a short modal description
     * @type {?|undefined}
     */
    RaveOptions.prototype.custom_description;
    /**
     * Link to your custom image
     * @type {?|undefined}
     */
    RaveOptions.prototype.custom_logo;
    /**
     * Text to be displayed as the title of the payment modal
     * @type {?|undefined}
     */
    RaveOptions.prototype.custom_title;
    /**
     * Amount to charge
     * @type {?}
     */
    RaveOptions.prototype.amount;
    /**
     * phone number of the customer
     * @type {?|undefined}
     */
    RaveOptions.prototype.customer_phone;
    /**
     * URL to redirect to when transaction is completed.
     * @type {?|undefined}
     */
    RaveOptions.prototype.redirect_url;
    /**
     * route country. Default NG
     * @type {?|undefined}
     */
    RaveOptions.prototype.country;
    /**
     * currency to charge the card in. Default "NGN"
     * @type {?}
     */
    RaveOptions.prototype.currency;
    /**
     * This allows you select the payment option you want for your users, possible values are card, account or both
     * @type {?|undefined}
     */
    RaveOptions.prototype.payment_method;
    /**
     * Text to be displayed on the Rave Checkout Button
     * @type {?|undefined}
     */
    RaveOptions.prototype.pay_button_text;
    /**
     * Unique transaction reference provided by the merchant
     * @type {?}
     */
    RaveOptions.prototype.txref;
    /**
     * This is a sha256 hash of your getpaidSetup values, it is used for passing secured values to the payment gateway
     * @type {?|undefined}
     */
    RaveOptions.prototype.integrity_hash;
    /**
     * The value to be passed for this is 1. This is useful when customer is using an opera browser, it
     * would load the payment modal on a new page.
     * @type {?|undefined}
     */
    RaveOptions.prototype.hosted_payment;
    /**
     * Any other custom data you wish to pass.
     * @type {?|undefined}
     */
    RaveOptions.prototype.meta;
    /**
     * Subaccounts to split payment with
     * https://developer.flutterwave.com/v2.0/docs/split-payment
     * @type {?|undefined}
     */
    RaveOptions.prototype.subaccount;
}
/**
 * @record
 */
export function PrivateRaveOptions() { }
if (false) {
    /**
     * A function to be called on successful card charge. User’s can always be redirected to a successful or
     * failed page supplied by the merchant here based on response
     * \@param response?: The server response
     * @type {?}
     */
    PrivateRaveOptions.prototype.callback;
    /**
     * A function to be called when the pay modal is closed.
     * @type {?}
     */
    PrivateRaveOptions.prototype.onclose;
    /**
     * A function to be called when payment is about to begin
     * @type {?}
     */
    PrivateRaveOptions.prototype.init;
}
var PaymentSetup = /** @class */ (function () {
    function PaymentSetup() {
    }
    return PaymentSetup;
}());
export { PaymentSetup };
if (false) {
    /** @type {?} */
    PaymentSetup.prototype.close;
}
/**
 * @record
 */
export function RaveRootOptions() { }
if (false) {
    /**
     * Your rave public key. You should use your test key for test mode and live key for live mode.
     * @type {?|undefined}
     */
    RaveRootOptions.prototype.key;
    /**
     * Used to determing what script to load. Set to false when using a test public key.
     * Default: `false`
     * @type {?|undefined}
     */
    RaveRootOptions.prototype.isTest;
}
/**
 * @record
 */
export function PaymentInstance() { }
if (false) {
    /**
     * Close the payment modal after payment has finished
     * @type {?}
     */
    PaymentInstance.prototype.close;
}
